### ISSUE TITLE
Are cumulative updates of % and %/y type incorrectly implemented?

### ISSUE THREAD

**Issue by ChaelKruip** [quintel/etengine#783]
# (Possible) Problem

@georgiasavvidou has pointed out to me that we possibly have an issue with update statements of the '%' and '%/y'-type. Focussing on the '%/y' update statements for the moment (but the '%' ones are equivalent), the future value is related to the user input (as a fraction) by

```
future_value = present_value * (1.0 + user_input)^number_of_years
```

where the `number_of_years` is the number of years between the present and the future (see [this piece of code for the real implementation](https://github.com/quintel/etengine/blob/master/app/models/gql/runtime/functions/update.rb#L113).

This implementation is correct for a single slider update on a single value (say 'A').
If the user touches two (or more) sliders of the '%/y'-type which affect the value of 'A', the engine will now multiply the effects of these sliders in the following way (as far as I understand it):

```
future_value = present_value * ((1.0 + user_input_1)^number_of_years) *  ((1.0 + user_input_2)^number_of_years) * ...
```

This results in counter-intuitive behaviour. Say that one slider is set to a growth of 'A' of 5%/y and the second slider is set to a decrease of 'A' of 5%/y. With a scenario for 2050, we would end up with:

```
A_future = A_present * 1.05^38 * 0.95^38 = A_present * 0.9092
```

whereas the user would (likely) expect the two effects to cancel out. Note that mixing one '%/y' and one '%' statement is not a problem.

@antw is it indeed the case that etengine multiplies the effect of sliders in the way described above?
# Affected parts of the ETM by '%/y' being used more than once

As far as I can see, the problem is limited to the following industry nodes:
- industry_other_metals_production: through sliders `industry_other_metals_production`, `industry_other_metals_process_electricity_efficiency` and `industry_other_metals_process_heat_useable_heat_efficiency`
- industry_useful_demand_for_chemical_electricity: through sliders `industry_useful_demand_for_chemical_electricity` and `industry_useful_demand_for_chemical_electricity_efficiency`
- industry_useful_demand_for_chemical_useable_heat: through sliders `industry_useful_demand_for_chemical_useable__heat` and `industry_useful_demand_for_chemical_useable__heat_efficiency`
- industry_useful_demand_for_other_electricity: through sliders `industry_useful_demand_for_other_electricity` and `industry_useful_demand_for_other_electricity_efficiency`
- industry_useful_demand_for_other_useable_heat: through sliders `industry_useful_demand_for_other_useable__heat` and `industry_useful_demand_for_other_useable__heat_efficiency`
# Affected parts of the ETM by '%' being used more than once

Similarly to '%/y'-type updates, multiplying the changes in '%' gives counter-intuitive effects:

Having a growth of 5% and an efficiency improvement of 5%, you would expect those to cancel out. The ETM would currently calculate:

```
A_future = A_present * 1.05 * 0.95 = A_present * 0.9975
```

were the user would expect

```
A_future = A_present * (1.0 + 0.05  - 0.05)  = A_present * 1.0 = A_present
```

I have not yet identified which parts of the ETM are affected by this. 
# Solution

The solution would be to sum all user inputs that affect a given quantity and apply the %/y only once (with the sum):

```
future_value = present_value * ((1.0 + user_input_1 + user_input_2 + ...)^number_of_years).
```

I can imagine that this might not be as simple as it seems given the current 'slider-by-slider' calculation method for updates.

Including @dennisschoenmakers @jorisberkhout @AlexanderWirtz

**Comment by dennisschoenmakers** [comment]
Good catch @georgiasavvidou !

**Comment by antw** [comment]
A good catch indeed!

> the user would (likely) expect the two effects to cancel out. Note that mixing one '%/y' and one '%' statement is not a problem. @antw is it indeed the case that etengine multiplies the effect of sliders in the way described above?

Yes. It [multiplies the `original_value` by the `user_input`](https://github.com/quintel/etengine/blob/e15ac4eda9500d970019c608020e508334a86a8e/app/models/gql/runtime/functions/update.rb#L145). The first time an input does this, `original_value` will be the same as `present_value`, but any subsequent calls will be the result changes made by the previous inputs.

> I can imagine that this might not be as simple as it seems given the current 'slider-by-slider' calculation method for updates.

I can see a way to do this by keeping track of the changes made by previous inputs. It's not trivial, but it should be possible.

**Comment by ChaelKruip** [comment]
One real life example is the efficiency improvement and the growth of electricity in the chemical industry:
![energietransitiemodel_-_je_eigen_onafhankelijke__volledige_en_op_feiten_gebaseerde_energiescenario_](https://cloud.githubusercontent.com/assets/1303760/9853340/64cf4582-5b03-11e5-8b27-a821781881af.png)
Setting both the sliders to 5%/y in a 2050 scenario results in almost 4 PJ difference with respect to the 'correct' approach. This corresponds to almost 10% of the value.

**Comment by AlexanderWirtz** [comment]
You may want to notify John. This is part of the Rli stuff

**Comment by dennisschoenmakers** [comment]
> This is part of the Rli stuff

We could at least fix the problem by adjusting the slider settings to take this effect into account, regardless of the necessity of a long-term fix.

**Comment by ChaelKruip** [comment]
> I have not yet identified which parts of the ETM are affected by this.

It seems to be limited to **appliances in households** which are affected both by efficiency sliders and behaviour sliders and lighting and appliances in buildings. 

The relevant sliders work with '%'-type updates.

**Comment by RobTerwel** [comment]
After I spotted a similar problem in the <a href="https://github.com/quintel/etsource/pull/1025https://github.com/quintel/etsource/pull/1025">transport sector<a/>, @jorisberkhout pointed me at this issue. I think the way I corrected those input statements there could also work here.

First of all, I think the solution proposed by @ChaelKruip  

> future_value = present_value * ((1.0 + user_input_1 + user_input_2 + ...)^number_of_years)

might be incorrect in some cases. This only works if user_input_1 etc. apply to the same variable, as @ChaelKruip rightfully noted. As I think it turns out, this is not always the case. For all '%y' industry statement for which we have statements that update useful demand ánd statements that update the efficiency, this probably does not work out.

I believe this is the case for the following reason: efficiency sliders should not change the useful demand, e.g. the heat produced. Efficiency sliders should ultimately change the demand of the specific production technology, while its output is still the same (i.e., the quantity required to meet the same useful demand). In other words, in the ETM in industry coal burners do not start generating more heat when their efficiencies increase - their fuel requirements and demand are just smaller.

If this reasoning is correct, we would just have to rewrite the update queries of the efficiency input statements such that they update the conversion attribute of the relevant production converters.
